src/raft.rs

```diff
@@ -2257,14 +2257,18 @@ impl<T: Storage> Raft<T> {
             "tag" => &self.tag,
         );
 
+        // Restore progress set and the learner flag.
         let next_idx = self.raft_log.last_index() + 1;
         let mut prs =
             ProgressSet::restore_snapmeta(meta, next_idx, self.max_inflight, &self.logger);
         prs.get_mut(self.id).unwrap().matched = next_idx - 1;
-        if self.is_learner && prs.configuration().voters().contains(&self.id) {
+        if prs.configuration().learners().contains(&self.id) {
+            self.is_learner = true;
+        } else if prs.configuration().voters().contains(&self.id) {
             self.is_learner = false;
         }
```

Comment:
What if it is neither in voter nor learner?

Reply:
So the peer will treat itself as a voter. It's ok because the leader knows it's removed.

Reply:
What if other peers do not know? Will it lead to split-brain?

Reply:
The peer is not in the configuration means it's removed, which must be commited by a majority of the cluster.